source/transfer_list.rst

```diff
@@ -146,6 +146,221 @@ The TE header is defined in :numref:`tab_te_header`.
+
+   #. Set `te.hdr_size` (`te_base_addr + 0x3`) to `0x8`.
+
+   #. Set `te.data_size` (`te_base_addr + 0x4`) to
+      `old_void_data_size - align8(new_data_size) - 0x8`.
+
+   #. Add the sum of the 8 bytes from `te_base_addr` to `te_base_addr + 0x8`
+      to `tl.checksum`.
+
+Adding a new TE with special data alignment requirement
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+Inputs:
+
+- `tl_base_addr`: Base address of the TL to add a TE to.
+- `new_tag_id`: ID number of the tag for the new TE.
+- `new_alignment`: The alignment boundary as a power of 2 that the data must be aligned to.
+- `new_data_size`: Size in bytes of the data to be encapsulated in the TE.
+- [data]: Data to be copied into the TE or generated on the fly.
+
+#. Calculate `alignment_mask` as `(1 << new_alignment) - 1`.
+
+#. If `(tl_base_addr + tl.size) & alignment_mask` is not 0, follow the steps in
+   `Adding a new TE`_ with the following different inputs (bypass the option to
+   overwrite an existing XFERLIST_VOID TE):
+
+   #. `tl_base_addr` remains the same
+
+   #. `new_tag_id` is 0 (XFERLIST_VOID)
+
+   #. `new_data_size` is `(1 << new_alignment) - ((tl_base_addr + tl.size) & alignment_mask) - 0x8`.
+
+   #. No data (i.e. just don't touch the bytes that form the data portion for this TE).
+
+#. Follow the steps in `Adding a new TE`_ with the original inputs (again bypass
+   the option to overwrite an existing XFERLIST_VOID TE).
+
+#. If `new_alignment` is larger than `tl.alignment`:
+
+   #. Subtract `tl.alignment` from `tl.checksum`.
+
+   #. Set `tl.alignment` to `new_alignment`.
+
+   #. Add `tl.alignment` to `tl.checksum`.
+
+Creating a TL
+^^^^^^^^^^^^^
+
+Inputs:
+
+- `tl_base_addr`: Base address where to place the new TL.
+- `available_size`: Available size in bytes to reserve for the TL after `tl_base_addr`.
+
+#. Check that `available_size` is larger than 16, otherwise abort.
+
+#. Set `tl.signature` (`tl_base_addr + 0x0`) to `0x6e_d0ff`.
+
+#. Set `tl.checksum` (`tl_base_addr + 0x4`) to 0 (for now).
+
+#. Set `tl.version` (`tl_base_addr + 0x5`) to 1.
+
+#. Set `tl.hdr_size` (`tl_base_addr + 0x6`) to 16.
+
+#. Set `tl.alignment` (`tl_base_addr + 0x7`) to 3.
+
+#. Set `tl.size` (`tl_base_addr + 0x8`) to 16.
+
+#. Set `tl.max_size` (`tl_base_addr + 0xc`) to `available_size`.
+
+#. Calculate the checksum as the sum of all bytes from `tl_base_addr` to
+   `tl_base_addr + 0x10`, and write the result to `tl.checksum`.
+
+Relocating a TL
+^^^^^^^^^^^^^^^
+
+Inputs:
+
+- `tl_base_addr`: Base address of the existing TL.
+- `target_base`: Base address of the target region to relocate into.
+- `target_size`: Total size in bytes of the target region to relocate into.
+
+#. Calculate `alignment_mask` as `(1 << tl.alignment) - 1`.
+
+#. Calculate the current `alignment_offset` as `tl_base_addr & alignment_mask`.
+
+#. Calculate `new_tl_base` as `(target_base & ~alignment_mask) + alignment_offset`.
```

Comment:
Would this work?

new_tl_base = ((target_base + alignment_mask) & ~alignment_mask) + alignment_offset`.

Reply:
I may be wrong, but should this be simply;
"new_tl_base = ((target_base + alignment_mask) & ~alignment_mask)" ?

Reply:
example:
 target_base = 0x1008
 tl.alignment = 0x20 -> alignment_mask = 0x1f
 ~alignment_mask = 0xffe0

 new_tl_base = (0x1008 + 0x1f) & 0xffe0 = 0x1027 & 0xffe0 = 0x1020
 
 we probably do not need the "alignment _offset" variable and the line 348 below.

Reply:
Agreed, this is simpler, thanks!

Reply:
Sorry, upon re-reading this I don't know what I was thinking here. This makes no sense. The `alignment_offset` is necessary because that's the only way the old `tl_base_addr` gets into this calculation at all. The new code here doesn't handle TL's that are not themselves aligned to the maximum alignment requirement correctly.

Uploaded PR #12 to fix this and revert to the original version of this PR for this section.